pkg/vsphere/toolbox/service.go

```diff
@@ -0,0 +1,210 @@
+}
+
+// Start initializes the RPC channels and starts a goroutine to listen for incoming RPC requests
+func (s *Service) Start() error {
+	err := s.in.Start()
+	if err != nil {
+		return err
+	}
+
+	err = s.out.Start()
+	if err != nil {
+		return err
+	}
+
+	ticker := time.NewTicker(s.Interval)
+
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+
+		for {
+			select {
+			case <-ticker.C:
+				_ = s.in.Send(nil) // POKE
+
+				request, _ := s.in.Receive()
+
+				if len(request) > 0 {
+					response := s.Dispatch(request)
+
+					_ = s.in.Send(response)
+				}
+			case <-s.stop:
+				ticker.Stop()
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// Stop cancels the RPC listener routine created via Start
+func (s *Service) Stop() {
+	s.stop <- struct{}{}
+
+	_ = s.in.Stop()
+	_ = s.out.Stop()
+}
+
+// Wait blocks until Start returns
+func (s *Service) Wait() {
+	s.wg.Wait()
+}
+
+// Handler is given the raw argument portion of an RPC request and returns a response
+type Handler func([]byte) ([]byte, error)
+
+// RegisterHandler for the given RPC name
+func (s *Service) RegisterHandler(name string, handler Handler) {
+	s.handlers[name] = handler
+}
+
+// Dispatch an incoming RPC request to a Handler
+func (s *Service) Dispatch(request []byte) []byte {
+	msg := bytes.SplitN(request, []byte{' '}, 2)
+	name := msg[0]
+
+	// Trim NULL byte terminator
+	if len(name) > 0 && name[len(name)-1] == 0 {
+		name = name[:len(name)-1]
+	}
+
+	handler, ok := s.handlers[string(name)]
+
+	if !ok {
+		log.Printf("unknown command: '%s'\n", name)
+		return []byte("Unknown Command")
+	}
+
+	var args []byte
+	if len(msg) == 2 {
+		args = msg[1]
+	}
+
+	response, err := handler(args)
+	if err == nil {
+		response = append([]byte("OK "), response...)
+	} else {
+		log.Printf("error calling %s: %s\n", name, err)
+		response = append([]byte("ERR "), response...)
+	}
+
+	return response
+}
+
+// Reset is the default Handler for reset requests
+func (s *Service) Reset([]byte) ([]byte, error) {
+	return []byte("ATR " + s.name), nil
+}
+
+// Ping is the default Handler for ping requests
+func (s *Service) Ping([]byte) ([]byte, error) {
+	return nil, nil
+}
+
+// SetOption is the default Handler for Set_Option requests
+func (s *Service) SetOption(args []byte) ([]byte, error) {
+	opts := bytes.SplitN(args, []byte{' '}, 2)
+	key := string(opts[0])
+	val := string(opts[1])
+
+	if Trace {
+		fmt.Fprintf(os.Stderr, "set option '%s'='%s'\n", key, val)
+	}
+
+	switch key {
+	case "broadcastIP": // TODO: const-ify
+		if val == "1" {
+			ip := s.PrimaryIP()
+			msg := fmt.Sprintf("info-set guestinfo.ip %s", ip)
+			return nil, s.out.Send([]byte(msg))
+		}
+	default:
+		// TODO: handle other options...
+	}
+
+	return nil, nil
+}
+
+// DefaultIP is used by default when responding to a Set_Option broadcastIP request
+// It can be overridden with the Service.PrimaryIP field
+func DefaultIP() string {
+	addrs, err := net.InterfaceAddrs()
+	if err == nil {
+		for _, addr := range addrs {
```

Comment:
tools/vmomi doesn't care what value we send back, so it is up to us to decide which IP would be the most useful to report.  There is also another RPC to report an array of NIC info, I haven't looked at yet - but in any case the IP value reported here is the most prominent in the UI and most commonly read via the API (e.g. govc vm.ip)